### PR TITLE
[WIP]1.add test_transactions test 2.fix generator account slicing 3.add ge…

### PR DESCRIPTION
**Pull request by y0sher**
…nkeypair func to eddsa

### CODE REVIEW

File: tests/test_bs.py [hunk 1]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+                print(pk)
+                assert test_account(pk)
+                ready+=1
+
+    assert ready == len(accounts)-1
+
+    end = time.time()
+    print("test took {:.3f} seconds ".format(end - start))
+
+    ## IF LOGEVITY
```
Comment (y0sher) [inline tests/test_bs.py]: TODO: add config param

File: tests/test_bs.py [hunk 2]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+            amount = random.randint(1, expected_balance(frm))
+        if gasprice is None:
+            gasprice = int(amount/10)
+        if gaslimit is None:
+            gaslimit = gasprice+1
+        txBytes = txGen.generate(to, accounts[frm]['nonce'], gaslimit, gasprice, amount)
+        accounts[frm]['nonce']+=1
+        data = '{"tx":'+ str(list(txBytes)) + '}'
+        out = api_call(pod_ip, data, submitapi, testconfig['namespace'])
+        print("submit transaction from {3} to {0} of {1} with gasprice {2}".format(to, amount, gasprice, frm))
```
Comment (beckmani) [inline tests/test_bs.py:441]: Move upwards. In case of failure you will not see this print

File: tests/test_bs.py [hunk 3]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+            print("nonce ok")
+            for pk in accounts:
+                if pk == tap:
+                    continue
+                print("checking account")
+                print(pk)
+                assert test_account(pk)
+                ready+=1
+
+    assert ready == len(accounts)-1
```
Comment (beckmani) [inline tests/test_bs.py]: Add assert string to describe what was wrong.
Also, why not put the assert inside test_account?
Something liks: "Expected amount:x, Existing amount: Y"

File: tests/test_bs.py [hunk 4]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+
+    for i in range(TEST_TXS2):
+        pod_ip, pod_name = random_node()
+        print("Sending tx from client: {0}/{1}".format(pod_name, pod_ip))
+        if i % 2 == 0:
+
+            # create new acc
+            acc, acc_info = random_acconut()
+            pub = new_account()
+            assert transfer(acc, pub)
```
Comment (beckmani) [inline tests/test_bs.py]: I would write a different test for transferring from newly created account

File: tests/test_bs.py [hunk 5]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+        else:
+            accfrom, accfrom_info = random_acconut()
+            accto, accto_info = random_acconut()
+            while accfrom == accto:
+                accfrom, accfrom_info = random_acconut()
+                accto, accto_info = random_acconut()
+
+            assert transfer(accfrom, accto)
```
Comment (beckmani) [inline tests/test_bs.py]: Add assert text

File: tests/test_bs.py (hunk 4 above)
Comment (beckmani) [inline tests/test_bs.py]: Add assert text

**Review by beckmani — COMMENTED** [review]
@y0sher , please follow the pep8 naming convention of variables (especially "lower_case_with_underscores" for multiwords)
https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: tests/test_bs.py [hunk 6]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+        print(out)
+        if str(accounts[acc]['nonce']) in out:
+            # check balance
+            data = '{"address":"' + str(acc) + '"}'
+            print(data)
+            out = api_call(pod_ip, data, balanceapi, testconfig['namespace'])
+            print(out)
+            balance = 0
+            if init_amount is not None:
+                balance += init_amount
```
Comment (beckmani) [inline tests/test_bs.py]: you can set "init_amount" to 0 (default) and simply add it to balance in line 459. - Then you will not need this if statement.

File: tests/test_bs.py [hunk 7]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+            balance = balance + expected_balance(acc)
+            if "{'value': '" + str(balance) + "'}" in out:
+                print( "{0}, balance ok ({1})".format(str(acc), out))
+                return True
+            return False
+        return False
+
+
+    # send tx to client via rpc
+    TEST_TXS = 10
```
Comment (beckmani) [inline tests/test_bs.py:483]: export this to config

File: tests/test_bs.py [hunk 8]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+    for i in range(TEST_TXS):
+        pod_ip, pod_name = random_node()
+        print(str("Sending tx from client: {0}/{1}").format(pod_name, pod_ip))
+        balance = tap_init_amount + expected_balance(tap)
+        if balance < 10:
```
Comment (beckmani) [inline tests/test_bs.py]: please explain in comment

File: tests/test_bs.py [hunk 9]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+            break
+        amount = random.randint(1, tap_init_amount + balance)
+        strpub = new_account()
+        assert transfer(tap, strpub, amount=amount)
```
Comment (beckmani) [inline tests/test_bs.py]: Add assert string

File: tests/test_bs.py [hunk 10]
```diff
@@ -388,6 +390,156 @@ def test_transaction(setup_network):
...
+        time.sleep(float(testconfig['client']['args']['layer-duration-sec']))
+        ready = 0
+        for pk in accounts:
+            if test_account(pk):
+                ready+=1
+
+        if ready == len(accounts)-1:
+            break
+
+    assert ready == len(accounts)-1
```
Comment (beckmani) [inline tests/test_bs.py]: Add assert string

**Review by beckmani — CHANGES_REQUESTED** [review]
See my comments inside

File: tests/test_bs.py (hunk 7 above)
Comment (y0sher) [reply tests/test_bs.py:483]: note that this is a specific test constant, I don't think that we need to export specific test-related constants to our global config. also this code is always run with the same constant config file so it won't really have an effect as if changing the value here or in config isn't much different.

File: tests/test_bs.py [hunk 11]
```diff
@@ -405,61 +407,71 @@ def test_transactions(setup_network):
     accounts = { tap: { "priv": "81c90dd832e18d1cf9758254327cb3135961af6688ac9c2a8c5d71f73acc5ce5", "nonce": 0, "send": [], "recv": [] } }
 
     def random_node():
-        rnd = random.randint(0, len(setup_network.clients.pods)-1)
-        return setup_network.clients.pods[rnd]['pod_ip'], setup_network.clients.pods[rnd]['name']
+        # rnd = random.randint(0, len(setup_network.clients.pods)-1)
```
Comment (beckmani) [inline tests/test_bs.py:410]: Delete?

**Comment by zalmen** [comment]
Was already merged with @noamnelke 's new tx flow

**Pull request closed without merging**